src/main/java/org/folio/tlib/postgres/impl/PgCqlQueryImpl.java

```diff
@@ -0,0 +1,315 @@
+  }
+
+  /**
+   * Convert CQL term to Postgres term - exact - without C style escapes in result.
+   * <p> Double backslash is converted to backslash. Postgres quotes (') are escaped.
+   * Otherwise things are passed through verbatim.
+   * </p>
+   * @param termNode termNode which includes term and relation.
+   * @return Postgres term without C style escapes.
+   */
+  static String cqlTermToPgTermExact(CQLTermNode termNode) {
+    String cqlTerm = termNode.getTerm();
+    StringBuilder pgTerm = new StringBuilder();
+    boolean backslash = false;
+    for (int i = 0; i < cqlTerm.length(); i++) {
+      char c = cqlTerm.charAt(i);
+      if (c == '\\' && backslash) {
+        backslash = false;
+      } else {
+        pgTerm.append(c);
+        if (c == '\'') {
+          pgTerm.append('\''); // important to avoid SQL injection
+        }
+        backslash = c == '\\';
+      }
+    }
+    return pgTerm.toString();
+  }
+
+  /**
+   * CQL full text term to Postgres term.
+   * @see <a href="https://www.postgresql.org/docs/13/sql-syntax-lexical.html">
+   *   String Constants section</a>
+   *
+   * <p>At this stage masking is unsupported and rejected.</p>
+   * @param termNode which includes term and relation.
+   * @return Postgres term.
+   */
+  static String cqlTermToPgTermFullText(CQLTermNode termNode) {
+    String cqlTerm = termNode.getTerm();
+    StringBuilder pgTerm = new StringBuilder();
+    boolean backslash = false;
+    for (int i = 0; i < cqlTerm.length(); i++) {
+      char c = cqlTerm.charAt(i);
+      // handle the CQL specials *, ?, ^, \\, rest are passed through as is
+      if (c == '*') {
+        if (!backslash) {
+          throw new IllegalArgumentException("Masking op * unsupported for: " + termNode.toCQL());
+        }
+      } else if (c == '?') {
+        if (!backslash) {
+          throw new IllegalArgumentException("Masking op ? unsupported for: " + termNode.toCQL());
+        }
+      } else if (c == '^') {
+        if (!backslash) {
+          throw new IllegalArgumentException("Anchor op ^ unsupported for: " + termNode.toCQL());
+        }
+      } else if (c != '\\' && backslash) {
+        pgTerm.append('\\');
+      }
+      if (c == '\\' && !backslash) {
+        backslash = true;
+      } else {
+        pgTerm.append(c);
+        if (c == '\'') {
+          pgTerm.append(c);
+        }
+        backslash = false;
+      }
+    }
+    return pgTerm.toString();
+  }
+
+  String handleTypeText(PgCqlField field, CQLTermNode termNode, boolean fullText) {
+    String s = handleNull(field, termNode);
+    if (s != null) {
+      return s;
+    }
+    String base = termNode.getRelation().getBase();
+    if (fullText) {
+      fullText = "=".equals(base) || "all".equals(base);
+    }
+    if (fullText) {
+      String pgTerm = cqlTermToPgTermFullText(termNode);
+      return "to_tsvector('" + language + "', " + field.getColumn() + ") @@ plainto_tsquery('"
+          + language + "', '" + pgTerm + "')";
```

Comment:
plainto_tsquery and phraseto_tsquery do not support right truncation.
Why don't you use to_tsquery and the existing well tested code that supports adj, all, any and right truncation?
https://github.com/folio-org/raml-module-builder/blob/v33.0.2/cql2pgjson/src/main/java/org/folio/cql2pgjson/util/Cql2SqlUtil.java#L260-L422
https://github.com/folio-org/raml-module-builder/blob/v33.0.2/cql2pgjson/src/test/java/org/folio/cql2pgjson/util/Cql2SqlUtilTest.java#L171-L479

Reply:
I don't like to copy so much code (the current implementation is much smaller).. But if need truncation and all, then there's no other way. It should be doable.